### PR TITLE
stream: process more than one write req per loop tick

### PR DESCRIPTION
**Pull request by ywave620**
Resolve https://github.com/libuv/libuv/issues/3668
Follow https://github.com/libuv/libuv/pull/3723 reviewer's suggestions, merge to v1.x rather than master.

### CODE REVIEW

File: src/unix/stream.c [hunk 1]
```diff
@@ -841,9 +841,16 @@ static void uv__write(uv_stream_t* stream) {
   QUEUE* q;
   uv_write_t* req;
   ssize_t n;
+  int count;
 
   assert(uv__stream_fd(stream) >= 0);
 
+  /* Prevent loop starvation when the consumer of this stream read as fast as
+   * (or faster than) we can write it. This `count` mechanism does not need to
+   * change even we switch to edge-triggered I/O.
```
Comment (vtjnash) [inline src/unix/stream.c:850]:
```suggestion
   * change even if we switch to edge-triggered I/O.
```

File: test/test-tcp-write-in-a-row.c [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+static uv_write_t big_write;
+
+/* 10 MB, which is large than the send buffer size and the recv buffer */
+static char data[1024 * 1024 * 10];
+
+static void close_cb(uv_handle_t *handle) {
+  close_cb_called++;
+}
+
+static void write_cb(uv_write_t *w, int status) {
```
Comment (vtjnash) [inline test/test-tcp-write-in-a-row.c:46]: style nit (for this file): `*` leans left
```suggestion
static void write_cb(uv_write_t* w, int status) {
```

File: test/test-tcp-write-in-a-row.c [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  ASSERT_EQ(0, uv_run(uv_default_loop(), UV_RUN_DEFAULT));
+
+  ASSERT_EQ(1, connect_cb_called);
+  ASSERT_EQ(3, close_cb_called);
+  ASSERT_EQ(1, connection_cb_called);
+  ASSERT_EQ(2, write_cb_called);
+
+  MAKE_VALGRIND_HAPPY();
+  return 0;
+}
```
Comment (vtjnash) [inline test/test-tcp-write-in-a-row.c:141]:
```suggestion
}
```

**Review by vtjnash — COMMENTED** [review]
LGTM

File: Makefile.am [hunk 4]
```diff
@@ -276,6 +276,7 @@ test_run_tests_SOURCES = test/blackhole-server.c \
                          test/test-tcp-writealot.c \
                          test/test-tcp-write-fail.c \
                          test/test-tcp-try-write.c \
+	                 test/test-tcp-write-in-a-row.c \
```
Comment (bnoordhuis) [inline Makefile.am:279]:
```suggestion
                         test/test-tcp-write-in-a-row.c \
```
(I think - indent error at any rate)

File: src/unix/stream.c [hunk 5]
```diff
@@ -862,10 +869,13 @@ static void uv__write(uv_stream_t* stream) {
       req->send_handle = NULL;
       if (uv__write_req_update(stream, req, n)) {
         uv__write_req_finish(req);
-        return;  /* TODO(bnoordhuis) Start trying to write the next request. */
+        if (count-- > 0) {
+          continue; /* Start trying to write the next request. */
+        }
```
Comment (bnoordhuis) [inline src/unix/stream.c:874]: Style:
```suggestion
        if (count-- > 0)
          continue;  /* Start trying to write the next request. */
```

**Review by bnoordhuis — APPROVED** [review]
Looks good modulo minor style nits. Thanks!

File: test/test-tcp-write-in-a-row.c (hunk 2 above)
Comment (santigimeno) [inline test/test-tcp-write-in-a-row.c:46]: Style nit: star leans left.

```suggestion
static void close_cb(uv_handle_t* handle) {
  close_cb_called++;
}

static void write_cb(uv_write_t* w, int status) {
```

File: test/test-tcp-write-in-a-row.c [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+  if (write_cb_called == 2) {
+    /* we are done */
+    uv_close((uv_handle_t*) &client, close_cb);
+    uv_close((uv_handle_t*) &incoming, close_cb);
+    uv_close((uv_handle_t*) &server, close_cb);
+  }
+}
+
+static void connect_cb(uv_connect_t *_, int status) {
```
Comment (santigimeno) [inline test/test-tcp-write-in-a-row.c:60]:
```suggestion
static void connect_cb(uv_connect_t* _, int status) {
```

File: test/test-tcp-write-in-a-row.c [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  /* fire a small write, which should be queued */
+  buf = uv_buf_init("A", 1);
+  r = uv_write(&big_write, (uv_stream_t*) &client, &buf, 1, write_cb);
+  ASSERT_EQ(0, r);
+
+  write_queue_size1 = uv_stream_get_write_queue_size((uv_stream_t*) &client);
+  ASSERT_EQ(write_queue_size1, write_queue_size0 + 1);
+}
+
+static void alloc_cb(uv_handle_t *handle, size_t size, uv_buf_t *buf) {
```
Comment (santigimeno) [inline test/test-tcp-write-in-a-row.c:86]:
```suggestion
static void alloc_cb(uv_handle_t* handle, size_t size, uv_buf_t* buf) {
```

File: test/test-tcp-write-in-a-row.c [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+  static char base[1024];
+
+  buf->base = base;
+  buf->len = sizeof(base);
+}
+
+static void read_cb(uv_stream_t *tcp, ssize_t nread, const uv_buf_t *buf) {}
```
Comment (santigimeno) [inline test/test-tcp-write-in-a-row.c:93]:
```suggestion
static void read_cb(uv_stream_t* tcp, ssize_t nread, const uv_buf_t* buf) {}
```

File: test/test-tcp-write-in-a-row.c [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+static void connection_cb(uv_stream_t *tcp, int status) {
```
Comment (santigimeno) [inline test/test-tcp-write-in-a-row.c:95]:
```suggestion
static void connection_cb(uv_stream_t* tcp, int status) {
```

**Review by santigimeno — APPROVED** [review]
Still LGTM with some style nits. Thanks!

**Comment by ywave620** [comment]
The test fail on `test/test-platform-output.c`. But I think it has nothing to do with this PR
```log
2022-08-30T10:35:51.1232498Z   uv_run_tests_a.vcxproj -> D:\a\libuv\libuv\build\Debug\uv_run_tests_a.exe
2022-08-30T10:35:51.1843273Z   Building Custom Rule D:/a/libuv/libuv/CMakeLists.txt
2022-08-30T10:35:51.2590199Z ##[group]Run build\\Debug\\uv_run_tests.exe platform_output
2022-08-30T10:35:51.2590648Z [36;1mbuild\\Debug\\uv_run_tests.exe platform_output[0m
2022-08-30T10:35:51.2618354Z shell: C:\Windows\system32\cmd.EXE /D /E:ON /V:OFF /S /C "CALL "{0}""
2022-08-30T10:35:51.2618645Z ##[endgroup]
2022-08-30T10:36:03.9587647Z not ok 1 - platform_output
2022-08-30T10:36:03.9588501Z # timeout
2022-08-30T10:36:03.9588865Z # Output from process `platform_output`:
2022-08-30T10:36:03.9589327Z # uv_get_process_title: Administrator: Windows PowerShell
2022-08-30T10:36:03.9589699Z # uv_cwd: D:\a\libuv\libuv
2022-08-30T10:36:03.9590071Z # uv_resident_set_memory: 7442432
2022-08-30T10:36:03.9590432Z # uv_uptime: 769.281000
2022-08-30T10:36:03.9590749Z # uv_getrusage:
2022-08-30T10:36:03.9591062Z #   user: 0 sec 0 microsec
2022-08-30T10:36:03.9591389Z #   system: 0 sec 0 microsec
2022-08-30T10:36:03.9591765Z #   page faults: 1934
2022-08-30T10:36:03.9592130Z #   maximum resident set size: 7356
2022-08-30T10:36:03.9592471Z # uv_available_parallelism: 2
2022-08-30T10:36:03.9592796Z # uv_cpu_info:
2022-08-30T10:36:03.9593168Z #   model: Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
2022-08-30T10:36:03.9593526Z #   speed: 2594
2022-08-30T10:36:03.9593804Z #   times.sys: 31656
2022-08-30T10:36:03.9594114Z #   times.user: 62281
2022-08-30T10:36:03.9594431Z #   times.idle: 675375
2022-08-30T10:36:03.9594719Z #   times.irq: 250
2022-08-30T10:36:03.9595015Z #   times.nice: 0
2022-08-30T10:36:03.9595377Z #   model: Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
2022-08-30T10:36:03.9595730Z #   speed: 2594
2022-08-30T10:36:03.9596009Z #   times.sys: 38750
2022-08-30T10:36:03.9596323Z #   times.user: 93015
2022-08-30T10:36:03.9596635Z #   times.idle: 637031
2022-08-30T10:36:03.9596924Z #   times.irq: 468
2022-08-30T10:36:03.9597227Z #   times.nice: 0
2022-08-30T10:36:05.9256909Z ##[error]Process completed with exit code 255.
```

**Comment by ywave620** [comment]
@bnoordhuis @bnoordhuis @vtjnash Shall we proceed to get this PR landed?

**Comment by bnoordhuis** [comment]
Sorry for the delay. Merged!

**Pull request merged**